src/Security/OAuthUserProvider.php

```diff
@@ -25,65 +27,72 @@
 final class OAuthUserProvider implements OAuthAwareUserProviderInterface
 {
     /**
-     * @var Registry
+     * @var RegistryInterface
      */
     private $doctrine;
 
-    public function __construct(Registry $doctrine)
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    public function __construct(RegistryInterface $doctrine, EventDispatcherInterface $eventDispatcher)
     {
-        $this->doctrine = $doctrine;
+        $this->doctrine        = $doctrine;
+        $this->eventDispatcher = $eventDispatcher;
     }
 
     public function loadUserByOAuthUserResponse(UserResponseInterface $response)
     {
-        $name     = $response->getResourceOwner()->getName();
-        $username = $response->getUsername();
-        $email    = $response->getEmail();
-        if (is_null($email)) {
-            throw new AccessDeniedException('No e-mailaddress available in this OAuth provider. Can\'t connect to it.');
+        if (is_null($response->getEmail())) {
+            throw new AccessDeniedException('No email address available in this OAuth provider. Can\'t connect to it.');
         }
 
-        $user = $this->loadDatabaseUser($name, $username, $email);
-        $this->updateUserRoles($user, $name);
+        $user = $this->loadDatabaseUser($response);
+        $this->updateUserRoles($user, $response->getResourceOwner()->getName());
 
         /** @var ObjectManager $objectManager */
-        $objectManager = $this->doctrine->getManagerForClass(User::class);
+        $objectManager = $this->doctrine->getManagerForClass(
+            $this->doctrine->getRepository(UserInterface::class)->getClassName()
+        );
         $objectManager->persist($user);
         $objectManager->flush();
 
         return $user;
     }
 
-    private function loadDatabaseUser(string $name, string $username, string $email): User
+    private function loadDatabaseUser(UserResponseInterface $response): UserInterface
     {
+        $name     = $response->getResourceOwner()->getName();
+        $username = $response->getUsername();
+
         /** @var UserRepository $repository */
         $repository = $this->doctrine->getRepository(UserInterface::class);
         $user       = $repository->findOneByOAuthUsername($name, $username);
 
-        if (is_null($user)) {
-            $user = $repository->findOneByEmail($email);
+        if (is_null($user) === false) {
```

Comment:
I think it should be tested that it is an instance of UserInterface, otherwise it will result in an error anyway, because of the return type of this method.